src/main/java/com/cloudbees/plugins/credentials/domains/CredentialSet.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015, Jacob Keller.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.plugins.credentials.domains;
+
+import com.cloudbees.plugins.credentials.Credentials;
+import com.cloudbees.plugins.credentials.CredentialsMatcher;
+import com.cloudbees.plugins.credentials.matchers.AnyOfMatcher;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.TreeMap;
+import java.util.Map;
+import java.util.List;
+
+/**
+ * Represents a mapping between {@link Domain}s and {@link Credentials},
+ * similar to {@link DomainCredentials}, but providing exact credentials for
+ * each domain.
+ */
+public class CredentialSet extends TreeMap<Domain, Credentials> {
```

Comment:
So I don't see it as a map between Domains and Credentials... rather a `Map<Set<DomainSpecification>,Credential>`. I also find inheriting from `TreeMap` as risky... for one the on-disk format should probably be `Map<Set<DomainSpecification>,String>` i.e. persisting the credential id rather than the credential itself...

And once you head down that road you actually probably want to just hide the map contract entirely and just give people a helper method... IOW you want to have them do

```
CredentialSet set = ...;
...
FooCredentials fooCreds = set.resolveCredentials(FooCredentials.class, context, auth, domainRequirements);
```

I suspect we want to flesh out some use cases as integrators would want first before trying to code this up